webcompat/helpers.py

```diff
@@ -36,7 +36,7 @@
                   '127.0.0.1', 'localhost')
 FIXTURES_PATH = os.getcwd() + '/tests/fixtures'
 STATIC_PATH = os.getcwd() + '/webcompat/static'
-JSON_MIME = 'application/json'
+JSON_MIME = 'application/vnd.github.v3.html+json'
```

Comment:
I think we probably don't want to request `html+json` for all JSON requests here. Perhaps we can add a new `HTML_JSON_MIME` or something, and use that where we need it?

(unless you've already verified we actually do need this for all JSON requests)

Reply:
Here's all the uses of `JSON_MIME` in this file:

https://github.com/webcompat/webcompat.com/blob/b2619d99a2ad85852ac47fa4c7d1fb24799efb18/webcompat/helpers.py#L190
https://github.com/webcompat/webcompat.com/blob/b2619d99a2ad85852ac47fa4c7d1fb24799efb18/webcompat/helpers.py#L203
https://github.com/webcompat/webcompat.com/blob/b2619d99a2ad85852ac47fa4c7d1fb24799efb18/webcompat/helpers.py#L373

So you're saying it's expected behavior to be requesting `vnd.github.v3.html+json` in all those places? Doesn't that mean we're sending that MIME on every `api_request`, for example? https://github.com/webcompat/webcompat.com/blob/b2619d99a2ad85852ac47fa4c7d1fb24799efb18/webcompat/helpers.py#L450

It's entirely possible it doesn't make a visible difference -- and the fact that you pass all tests is probably indicative of that. What I'm wondering is if we should only request `vnd.github.v3.html+json` when we actually want an html response.

Reply:
@marimeireles maybe this *is* what we want, don't take my questions to mean you're doing something wrong :). I guess my question is, is there a situation where we aren't requesting HTML from GitHub, just plain JSON? Or is it all HTML now?

Reply:
Here's one: 

![screen shot 2018-12-21 at 9 52 34 am](https://user-images.githubusercontent.com/67283/50350902-2b8b4180-0506-11e9-821c-c89f16342108.png)

Perhaps we need @karlcow to chime in on this. I'm not sure what the consequences are for this change (claiming `application/json` is `application/vnd.github.v3.html+json` when it isn't). I think we're breaking HTTP semantics by doing so.

Reply:
>So you're saying it's expected behavior to be requesting vnd.github.v3.html+json in all those places?

Yes.

>I guess my question is, is there a situation where we aren't requesting HTML from GitHub, just plain JSON? Or is it all HTML now?

It's still a JSON, the same JSON we had before. I made only small changes on the JSONs to replace markdown for HTML. Here is an example of the kind of changes I did:

```
removed: "body": "\n**URL**: www.mmmmm.com\n**Browser**: Firefox\n**Version**: 30\n\nThe site asks me to upgrade",
added: "body_html": "<p><strong>URL</strong>: <a href=\"http://www.mmmmm.com\" rel=\"nofollow\">www.mmmmm.com</a><br> <strong>Browser</strong>: Firefox<br> <strong>Version</strong>: 30</p> <p>The site asks me to upgrade</p>",
```

The only part where we're still using markdown-it is for rendering the comments that the user writes on the website, because we need to send them as markdown to github and also because immediately after the user sends a comment we show the rendered comment on the webpage (without making any request to Github), so we need the lib to do that for us. 

But when we fetch the files from github we're also fetching the comments as HTML, everything we fetch from Github is HTML and IMO it's not necessary to use `application/json` anymore. If you think it's mandatory to use it somewhere because of some reason I didn't think of then I'd have to change the code because as far as I'm concerned the program will only work with JSONs with HTML from now on, instead of JSONs with markdown.

> maybe this is what we want, don't take my questions to mean you're doing something wrong :)

Okay :)

Reply:
>Perhaps we need @karlcow to chime in on this. I'm not sure what the consequences are for this change (claiming application/json is application/vnd.github.v3.html+json when it isn't). I think we're breaking HTTP semantics by doing so.

Me neither... We can wait for Karl, then. :)
According to [this](https://stackoverflow.com/a/20620606/10756713) answer on stackoverflow it doesn't change anything.
If we want to make things right... I think I'd just have to find how the `/webcompat/static/js/dist/issues.js` is generated and change the part that generates this:

```
    $.ajax({
      contentType: "application/json",
      data: JSON.stringify({
        milestone: statusObject.id,
        state: statusObject.state
      }),
      type: "PATCH",
      url: "/api/issues/" + $("main").data("issueNumber") + "/edit",
      success: _.bind(function() {
        var currentMilestone = _.find(this.get("milestones"), function(status) {
          return status.id === statusObject.id;
        });

        this.set("milestone", currentMilestone);
        this.issueModel.set("milestone", currentMilestone.name);
      }, this),
```

@karlcow am I in the right direction here?

Reply:
> I think I'd just have to find how the `/webcompat/static/js/dist/issues.js` is generated

See https://github.com/webcompat/webcompat.com/blob/master/grunt-tasks/concat.js#L39

> and change the part that generates this

Not sure what you're thinking, but yes, milestone API requests and label API requests should be sending a non-HTML mimetype. I was thinking we could extend `api_request` to take a mime override argument for the cases where we are requesting non-`application/json`.

Reply:
> milestone API requests and label API requests should be sending a non-HTML mimetype

Why do the milestones and labels can't request for a HTML?
The JSON file looks exactly the same, except for the markdown parts.
The only issue that I see is the one pointed [here](https://github.com/webcompat/webcompat.com/pull/2731#discussion_r243620137) about the content-type being wrong.
Sorry for insisting, but I really didn't understand.

> I was thinking we could extend api_request to take a mime override argument for the cases where we are requesting non-application/json.

Seems like a good solution, if we'll have to fetch application/json.

Reply:
Here's an example of requesting the labels as `html+json`:

```
🐓 http --print=hH GET https://api.github.com/repos/webcompat/web-bugs/labels **Accept:'application/vnd.github.v3.html+json'**
GET /repos/webcompat/web-bugs/labels HTTP/1.1
Accept: application/vnd.github.v3.html+json
Accept-Encoding: gzip, deflate
Connection: keep-alive
Host: api.github.com
User-Agent: HTTPie/1.0.0

HTTP/1.1 200 OK
Access-Control-Allow-Origin: *
Access-Control-Expose-Headers: ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type
Cache-Control: public, max-age=60, s-maxage=60
Content-Encoding: gzip
Content-Security-Policy: default-src 'none'
**Content-Type: application/json; charset=utf-8**
Date: Fri, 21 Dec 2018 21:35:28 GMT
ETag: W/"0c4c97b020aa6958117b5d8f3cad9bf6"
Link: <https://api.github.com/repositories/17914657/labels?page=2>; rel="next", <https://api.github.com/repositories/17914657/labels?page=5>; rel="last"
Referrer-Policy: origin-when-cross-origin, strict-origin-when-cross-origin
Server: GitHub.com
Status: 200 OK
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
Transfer-Encoding: chunked
Vary: Accept
X-Content-Type-Options: nosniff
X-Frame-Options: deny
X-GitHub-Media-Type: github.v3; param=html; format=json
X-GitHub-Request-Id: CD59:4747:1C34DB6:476FBBA:5C1D5C9F
X-RateLimit-Limit: 60
X-RateLimit-Remaining: 46
X-RateLimit-Reset: 1545431394
X-XSS-Protection: 1; mode=block
```

We requested `...html+json`, but GitHub just sent us back `application/json`, presumably because today that's all they support. I think it's problematic for the server to claim the content type is `...html+json`, because it isn't.

> Why do the milestones and labels can't request for a HTML?

They can, but a follow-up question is *why* would we request `...html+json` for labels or milestones? If one day GitHub sent back an actual HTML+JSON response, and we expect just plain JSON, that would probably break parts of our app. The fact that today they return `application/json` today is useful, but we can't rely on that behavior, especially if we're explicitly requesting a different content type

I'm also concerned about breaking HTTP semantics by **claiming** we're returning `application/vnd.github.v3.html+json` in all API requests, when in fact we're not -- we've just hard-coded that value. I suppose we've just been lucky that we've only requested `application/json` in the past, so hardcoding wasn't an obvious issue.

(and yes, this feels very theoretical and who cares, etc., but I think it's worth doing it right because it might lead to bugs or confusion in the future, in ways we can't predict 😸 ).

I think a path forward here is to be explicit about what we're requesting (sensible defaults make sense, with a way to override), and for responses proxy the **actual** `Content-Type` header value returned from GitHub.

Reply:
>but we can't rely on that behavior, especially if we're explicitly requesting a different content type

I see your point and I totally agree.

> (and yes, this feels very theoretical and who cares, etc., but I think it's worth doing it right because it might lead to bugs or confusion in the future, in ways we can't predict 😸 ).

I think it's super important to write a good code so we don't struggle with it in the future. We'll probably save time doing thins the right way. :)

Reply:
Thank you a lot for the discussion guys.
I'll start implementing these changes.

Reply:
I was reading more about the use of html+json and I learned that "The suffix "+json" MAY be used with any media type whose representation follows that established for "application/json"." [according](https://tools.ietf.org/html/rfc6839#section-3.1) to IETF, in other words a document with `vnd.github.v3.html+json` MIME type is an `application/json` document. 

Besides that Github says that every time there is a comment when we make this  `application/vnd.github.VERSION.html+json` request it will return the HTML version, but that will *only* happen for comments. Github wouldn't start sending HTML when we request a label because it'd mean that they're breaking their API. The version 3 that we use is stable.

So I think the ideal thing would be to always return `application/json` in our responses since `vnd.github.v3.html+json` is a Github vendor MIME type that isn't meaningful to our front end.

Reply:
Okay Karl, you convinced me with you wisdom. Thank you and Mike for being so patient, you're super nice. :) 

I'm sorry for insisting so much. I've wrote a patch, gonna ping you and Mike there to check if everything is right.